### ISSUE TITLE
UCX does not work with OpenMPI 4.0.5

### ISSUE THREAD

**Issue by matthiasdiener** [UIUC-PPL/charm#3079]
See https://github.com/UIUC-PPL/charm/pull/3064.

4.0.4 is fine.

**Comment by minitu** [comment]
Need to explore OpenPMIx as an alternative to using PMIx included in OpenMPI (as it is no longer included as of OpenMPI 4.0.5).

**Comment by minitu** [comment]
Seems to work fine on OLCF Summit, steps to reproduce:

- UCX
- OpenPMIx
    - Requires libevent
        - wget https://github.com/libevent/libevent/releases/download/release-2.1.12-stable/libevent-2.1.12-stable.tar.gz
        - cd libevent-2.1.12-stable
        - mkdir build && mkdir install
        - cd build
        - ../configure --prefix=$HOME/libevent-2.1.12-stable/install
        - make -j && make install
    - wget https://github.com/openpmix/openpmix/releases/download/v3.1.5/pmix-3.1.5.tar.gz
    - cd pmix-3.1.5
    - mkdir build && mkdir install
    - cd build
    - ../configure --prefix=$HOME/pmix-3.1.5/install --with-libevent=$HOME/libevent-2.1.12-stable/install
    - make -j && make install
- Charm++
    - ./buildold charm++ ucx-linux-ppc64le smp ompipmix -j --with-production --basedir=$HOME/pmix-3.1.5/install

**Comment by minitu** [comment]
Charm++ build first fails with
```
../bin/charmc  -optimize -production  -o ../lib/libconv-partition.a custom_partitioner.o set_partition_params.o
ar: creating ../lib/libconv-partition.a
../bin/charmc  -optimize -production  -o ../lib/libconv-ldb.a topology.o generate.o edgelist.o
ar: creating ../lib/libconv-ldb.a
../bin/charmc  -optimize -production  -c pup_f.f90
gfortran: warning: ../bin/../include: linker input file unused because linking not done
../bin/charmc  -optimize -production  -cpmod ../include pupmod.M
cp: cannot stat 'PUPMOD': No such file or directory
Fatal Error by charmc in directory /ccs/home/jchoi/work/charm-github/tmp
   Command cp -p PUPMOD ../include/ returned error code 1
charmc exiting...
make: *** [pup_f.o] Error 1
```
but error goes away when `make charm++` is run again in `tmp`.

**Comment by matthiasdiener** [comment]
> Charm++ build first fails with
> 
> ```
> ../bin/charmc  -optimize -production  -o ../lib/libconv-partition.a custom_partitioner.o set_partition_params.o
> ar: creating ../lib/libconv-partition.a
> ../bin/charmc  -optimize -production  -o ../lib/libconv-ldb.a topology.o generate.o edgelist.o
> ar: creating ../lib/libconv-ldb.a
> ../bin/charmc  -optimize -production  -c pup_f.f90
> gfortran: warning: ../bin/../include: linker input file unused because linking not done
> ../bin/charmc  -optimize -production  -cpmod ../include pupmod.M
> cp: cannot stat 'PUPMOD': No such file or directory
> Fatal Error by charmc in directory /ccs/home/jchoi/work/charm-github/tmp
>    Command cp -p PUPMOD ../include/ returned error code 1
> charmc exiting...
> make: *** [pup_f.o] Error 1
> ```
> 
> but error goes away when `make charm++` is run again in `tmp`.

Is it possible that there is an error before that? This looks normal.

**Comment by minitu** [comment]
The build issue was fixed with #3150. Will create a PR to add documentation to use OpenPMIx for UCX builds.

**Comment by minitu** [comment]
Link error occurs on Lassen, need to remove `-lopen-pal` and `-lopen-rte` from `src/arch/ucx/conv-mach-ompipmix.sh` (or create a new `conv-mach-pmix.sh`)
